vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,524 @@
+                        if !bat_found {
+                            bat_found = true;
+                            bat_entry = Some(entry);
+                            continue;
+                        }
+                        return Err(Error::DuplicateBATEntry);
+                    }
+
+                    if entry.guid == MDR_GUID {
+                        if !mdr_found {
+                            mdr_found = true;
+                            mdr_entry = Some(entry);
+                            continue;
+                        }
+                        return Err(Error::DuplicateMDREntry);
+                    }
+
+                    if (entry.required & REGION_ENTRY_REQUIRED) == 1 {
+                        // This implementation doesn't recognized this field.
+                        // Therefore, accoding to the spec, we are throwing an error.
+                        return Err(Error::UnrecognizedRegionEntry);
+                    }
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+        }
+
+        if !bat_found || !mdr_found {
+            region_entries.clear();
+            return Err(Error::RegionEntryCollectionFailed);
+        }
+
+        Ok(RegionInfo {
+            bat_entry,
+            mdr_entry,
+            region_entries,
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct RegionEntry {
+    start: u64,
+    end: u64,
+}
+
+enum HeaderNo {
+    First,
+    Second,
+}
+
+/// Contains the information from the header of a VHDx file
+#[derive(Clone, Debug)]
+pub struct VhdxHeader {
+    file_type_identifier: FileTypeIdentifier,
+    header_1: Header,
+    header_2: Header,
+    region_table_1: RegionTableHeader,
+    region_table_2: RegionTableHeader,
+}
+
+impl VhdxHeader {
+    /// Creates a VhdxHeader from a reference to a file
+    pub fn new(f: &mut File, guid: u128) -> Result<VhdxHeader> {
+        let file_type_identifier: FileTypeIdentifier = FileTypeIdentifier::new(f)?;
+        let header_1 = Header::new(f, HEADER1_START);
+        let header_2 = Header::new(f, HEADER2_START);
+        let (header_1, header_2) = match VhdxHeader::update_headers(f, header_1, header_2, guid) {
+            Ok(h) => h,
+            Err(e) => return Err(e),
+        };
+
+        Ok(VhdxHeader {
+            file_type_identifier,
+            header_1,
+            header_2,
+            region_table_1: RegionTableHeader::new(f, REGION1_START)?,
+            region_table_2: RegionTableHeader::new(f, REGION2_START)?,
+        })
+    }
+
+    /// Identify the current header and return both headers along with an integer indicating the current header
+    fn current_header(
+        header_1: Result<Header>,
+        header_2: Result<Header>,
+    ) -> Result<(HeaderNo, Header)> {
+        let mut header1_seq_num: u64 = 0;
+        let mut header2_seq_num: u64 = 0;
+        let mut no_valid_header: bool = true;
```

Comment:
Nit: It'd be more natural to call it `valid_hdr_found`, therefore inverting the logic of the boolean.